Feature.shapes

nick-chura opened the pull request:
This is the first version of the "shapes" macro library. It comes with an html help file.

The purpose is to simplify the creation of basic shapes with labels, using the existing showasciisvg macro in IMathAS.

drlippman commented:
Generally looks great!  A few questions:

1) in the basic angles and sides options for draw_triangle, is there a way to give it to always draw the triangle in the same orientation, like with right angle in the lower left corner?

2) for the "altitude", is there a way to label the altitude length itself?

3) Is there a way to control the size of the final diagram?  It seems to use the default 300x300.

As a bit of a random comment, I notice that often the resulting graph ends up having a lot of blank space.  I don't think that's a big deal, but it could make some diagrams harder to read, if the actual shape doesn't get the full space available.

nick-chura commented:
1. I added the "horizontal" option to have the longest side on the bottom horizontal, but that could be changed for right triangles. Are you thinking primarily of right triangles?
2. I considered adding a label for altitudes, but since that was the last option I wrote, I maybe just forgot. It would be easy to add, but I don't know how well I could optimize the label placement. I'll see what I can come up with.
3. Oh, yes. I kept the size at 300x300 since any smaller and the labels tended to interfere with each other. I can add an argument for "size,length,width".
4. Is the blank space thing specific to certain of the shapes?

drlippman commented:
1.  Yeah, mainly for right triangles.  For some reason it feels like it's common to see right triangles drawn with a horizontal leg.

4. The place I noticed it in particular was on the circle segment.  This really isn't a big deal.

nick-chura commented:
I've made some changes.

1. Triangles and polygons are now drawn in a "standard" position by default. [Vertices labeled counterclockwise from bottom-left, with bottom edge drawn horizontally.] Removed the "horizontal" option, added a "rotate" option for random rotation of the shape.
2. Added a "size" option to change the size of the image.
3. Fixed the image size for draw_circlesector.
4. Added a draw_prismcubes function.

drlippman commented on assessment/libs/shapes.php:
```diff
@@ -0,0 +1,1465 @@
+  
+  //number of sides
+  $n = $argsArray[0][0];
+  if (!is_numeric($n) || abs($n-round($n))>1E-9 || $n < 3) {
```
Small request: can you add a max value for $n too, to prevent the loops from getting too outrageous?  Maybe a max of 50 sides or something?

drlippman commented on assessment/libs/shapes.php:
```diff
@@ -0,0 +1,1465 @@
+    $args = $args . "line([$i,0],[$i,$height]);";
+    $args = $args . "line([$i,$height],[$i+$depth/sqrt(2),$height+$depth/sqrt(2)]);";
+  }
+  for ($i=0;$i<($depth+1);$i++) {
```
These three loops seem like they have the potential to generate a LOT of lines, if height, length, or depth was large.  Could we add a max for these?

drlippman commented:
Looking good.  I added a couple comments on places where it'd be good to a max limit on the parameter to prevent excessive looping.

nick-chura commented:
I fixed those things. Let me know if there's anything else you'd like changed.
